### ISSUE TITLE
Support minifying css output from sass compiler

### ISSUE THREAD

**Issue by vaadin-bot** [vaadin/sass-compiler#93]
**Originally by @Artur-**

___
There should be a tool included in Vaadin that takes care of minifying css. This tool should be used automatically to minify the output of the sass compiler. 

Running reindeer/styles.css through codebeautifier.com reduces the file size with 20% (10% gzipped)

___
 Imported from https://dev.vaadin.com/ issue #10812

**Comment by vaadin-bot** [comment]
**Originally by @jouni**

---

This should be the similar to what Ruby Sass [Output style](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#output_style) provides (the :compact options should be the initial target).

**Comment by vaadin-bot** [comment]
**Originally by @jouni**

---

Related: #14600

**Comment by vaadin-bot** [comment]
**Originally by @Legioth**

---

Two feasible options for doing this could be https://jawr.java.net/ or http://yui.github.io/yuicompressor/

**Comment by vaadin-bot** [comment]
**Originally by _ttran_**

---

This is what I'm currently using:

```
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>exec-maven-plugin</artifactId>
    <version>1.3.2</version>
    <executions>
        <execution>
            <id>conversion</id>
            <phase>generate-resources</phase>
            <configuration>
                <classpathScope>compile</classpathScope>
                <mainClass>com.genix.platform4.commons.sass.SassCompilerEx</mainClass>
                <arguments>
                    <argument>src/main/webapp/VAADIN/themes/protrack/styles.scss</argument>
                    <argument>src/main/webapp/VAADIN/themes/protrack/styles.full.css</argument>
                </arguments>
            </configuration>
            <goals>
                <goal>java</goal>
            </goals>
        </execution>
        <execution>
            <id>minify</id>
            <phase>generate-resources</phase>
            <configuration>
                <includePluginDependencies>true</includePluginDependencies>
                <mainClass>com.yahoo.platform.yui.compressor.YUICompressor</mainClass>
                <arguments>
                    <argument>--line-break</argument>
                    <argument>10</argument>
                    <argument>-o</argument>
                    <argument>src/main/webapp/VAADIN/themes/protrack/styles.css</argument>
                    <argument>src/main/webapp/VAADIN/themes/protrack/styles.full.css</argument>
                </arguments>
            </configuration>
            <goals>
                <goal>java</goal>
            </goals>
        </execution>
    </executions>
    <dependencies>
        <dependency>
            <groupId>com.yahoo.platform.yui</groupId>
            <artifactId>yuicompressor</artifactId>
            <version>2.4.7</version>
        </dependency>
    </dependencies>
</plugin>
```

**Comment by vaadin-bot** [comment]
**Originally by @thinwire**

---

There's now a patch that pulls in YUICompressor as a dependency via Maven, and is used by default to compress the final CSS output unless `-compress:false` is specified on the command line.
Since this change both alters the tool's default behavior and dependencies, as well as introducing better (and thus, different) command line argument handling, this change will at the earliest be seen with the 7.5 and later series of Vaadin.

https://dev.vaadin.com/review/#/c/4887/

**Comment by vaadin-bot** [comment]
**Originally by @hesara**

---

The option name should probably be "minify" or similar rather than "compress", which is reserved for gzip compression (#17773).

**Comment by vaadin-bot** [comment]
**Originally by @hesara**

---

Renaming parameter

**Comment by vaadin-bot** [comment]
**Originally by @hesara**

---

Renamed parameter in https://dev.vaadin.com/review/#/c/11581/

**Comment by vaadin-bot** [comment]
**Originally by _artamonov_**

---

Beware! Do not use yuicompressor 2.4.8, it produces wrong CSS and window close animations in IE does not work, any VOverlay cannot close.

**Comment by vaadin-bot** [comment]
**Originally by _artamonov_**

---

Also check 2.4.7 with Safari 9. Version 2.4.7 produces invalid animation CSS that leads to broken hide animation for overlays

**Comment by vaadin-bot** [comment]
**Originally by @hesara**

---

Disabled minifying by default in https://dev.vaadin.com/review/#/c/12567/ . It can be activated with -minify:true for themes that do support it. it is recommended to use GZIP compression instead of minifying - see also #17773 for pre-compression at theme compilation time.

**Comment by mvysny** [comment]
From my experience minify is not really needed. I tried to minify a CSS produced by compilation of a lightly modified Vaadin Valo 7.7.6 theme:

1. original CSS size: 300kb
2. minified CSS size: 280kb
3. gzip-compressed CSS size: 32kb
4. gzip-compressed minified CSS size: 29kb

So, you save nearly 90% by gzip-compressing of the theme, but you only save 3kb by further minifying the gzipped CSS. I have posted a small tutorial on how to enable compressing at #297
